src/mpid/ch4/include/mpidpost.h

```diff
@@ -95,7 +95,61 @@ MPL_STATIC_INLINE_PREFIX int MPID_Waitall(int count, MPIR_Request * request_ptrs
 
 MPL_STATIC_INLINE_PREFIX int MPID_Wait(MPIR_Request * request_ptr, MPI_Status * status)
 {
-    return MPIR_Wait_impl(request_ptr, status);
+    const int single_threaded = !MPIR_ThreadInfo.isThreaded;
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_Thread_sync_t *sync = NULL;
+
+    if (request_ptr == NULL)
+        return MPI_SUCCESS;
+
+    if (likely(single_threaded)) {
```

Comment:
Why aren't we just calling `MPIR_wait_impl` at this point? This code looks the same.

Reply:
Unfortunately the compiler insert a function call and adds overhead to the latency-critical path. So leaving it here save some cycles in my test.

Reply:
What compiler/configuration? We just spent the effort to remove all these duplicate progress loops. I'd rather not add one back in so soon.